pkg/gc/config.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"strconv"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	ConfigName = "config-gc"
+)
+
+type Config struct {
+	// Delay duration after a revision create before considering it for GC
+	StaleRevisionCreateDelay time.Duration
+	// Timeout since a revision lastPinned before it should be GC'd
+	StaleRevisionTimeout time.Duration
+	// Minimum number of generations of revisions to keep before considering for GC
+	StaleRevisionMinimumGenerations int64
+	// Minimum staleness duration before updating lastPinned
+	StaleRevisionLastpinnedDebounce time.Duration
+}
+
+func NewConfigFromConfigMap(configMap *corev1.ConfigMap) (*Config, error) {
+	c := Config{}
+
+	for _, dur := range []struct {
+		key          string
+		field        *time.Duration
+		defaultValue time.Duration
+	}{{
+		key:          "stale-revision-create-delay",
+		field:        &c.StaleRevisionCreateDelay,
+		defaultValue: 5 * time.Minute,
+	}, {
+		key:          "stale-revision-timeout",
+		field:        &c.StaleRevisionTimeout,
+		defaultValue: 5 * time.Minute,
```

Comment:
These should default to like a day or so.  I am worried about how this will interact with our 10 hour resync change with such low values.

Reply:
I had a question about this above which GH is doing a *great* job at hiding: https://github.com/knative/serving/pull/2000#discussion_r224246345

Reply:
At least for me, that comment link doesn't seem to work. Perhaps ask the question again?

Reply:
Related to the comment from @mattmoor above, am I reading the code correctly that we'll only update lastPinned on the Revision once every resync period? That resync used to be 30 seconds (I believe?) and is now 10 hours. So that means that we'll need far longer defaults here and in the configmap. And, it means in practice that revision's won't be garbage collected "quickly" with the longer resync period.

Reply:
pasting my above Q since I fail at using GH :)

This is making me wonder if relying on resync period for writing out lastPinned is the correct design: StaleRevisionTimeout will need to be longer than the controller resync period to avoid GCing pinned revisions. ISTM a user should be able to bump resync period up as long as they like and only suffer resources remaining stale for longer periods, not failing behavior. Maybe we should have some kind of goroutine / controller which writes out lastPinned periodically independent of resyncPeriod?

Reply:
Alternatively, I could make sure that StaleRevisionTimeout is always longer than syncperiod. IMO this isnt great and should probably come with a todo to make GC follow its own sync period but could be a good start.